program-runtime/src/invoke_context.rs

```diff
@@ -148,6 +194,7 @@ pub struct InvokeContext<'a> {
     compute_budget: ComputeBudget,
     current_compute_budget: ComputeBudget,
     compute_meter: Rc<RefCell<ComputeMeter>>,
+    accounts_data_meter: Rc<RefCell<AccountsDataMeter>>,
```

Comment:
Right now I have `accounts_data_meter` as an `Rc<RefCell<>>` because I need interior mutability. I'm still trying to figure out where I need to track accounts data len changes (basically calling `AccountsDataMeter::consume()`, which takes a `&mut self`).

So far I have two main spots I'm thinking about:

1. In `system_instruction_processor::allocate()`, and functions like it

In this case, `InvokeContext` is not mutable, so the RefCell is needed (-or- I pass InvokeContext as `&mut` all the way down). Refer to this example impl:
https://github.com/solana-labs/solana/blob/18e4638331b2d3ec9b3cb39a3364376e603c6d85/runtime/src/system_instruction_processor.rs#L108-L111

I know that `allocate()` is not sufficient to capture all the locations to monitor though.


2. In `InvokeContext::verify()` and `verify_and_update()`

In this case I have `&mut self`, so I'm good and don't need the RefCell. However, is this sufficient to capture all changes to accounts data len? If so, that would be great!

https://github.com/brooksprumo/solana/blob/5d7c3e5a35f6c8f9076fe43c5c41b42f22697095/program-runtime/src/invoke_context.rs#L450-L453

and then down on line 464 I could remove the `.borrow_mut()`.

So, I guess @Lichtso , do you know if (2) will work? Also, please let me know if there are questions, as I don't know how much context you already have (or what additionally is useful). TIA!

Reply:
One of the challenges here is to not double count.  Doing it in #1 I think will double count in #2 for CPI calls

Reply:
Doing it in #2 only is probably sufficient, the only issue with that is the check is after the system program has physically allocated the memory, which means we would want to add a 10mb buffer zone to the budget.  Catching it before the physical allocation would be ideal.

@brooksprumo Why not mirror what we do with `ComputeMeter` which is `Rc<RefCell`?

Reply:
> One of the challenges here is to not double count. Doing it in #1 I think will double count in #2 for CPI calls

Gotcha. I wouldn't do both #1 and #2.

> The check is after the system program has physically allocated the memory, which means we would want to add a 10mb buffer zone to the budget.

An additional 10 MB would not be an issue, so I'm not worried about that. We were initially thinking 1 TB instead of 128 GB, so it's already very conservative.

> Catching it before the physical allocation would be ideal.

I agree! Is there a way to do that cleanly though?

> Why not mirror what we do with ComputeMeter which is Rc<RefCell?

Currently `AccountsDataMeter` is a `Rc<RefCell<>>`. Since @Lichtso said that the `Rc<RefCell<>>` isn't the best, if I can do option 2 then I can _remove_ the `Rc<RefCell<>>`, since the uses of `InvokeContext::accounts_data_meter` will all be within an `impl InvokeContext` function with `&mut self`.

Reply:
I would mirror `ComputeMeter` for now, we can change both later if need be.  We could probably return a total changed data length in verify and update the alloc meter there

Reply:
Sounds good! So I'll leave this PR as-is, since `AccountsDataMeter` does mirror `ComputeMeter`.

For my next PR I'll update the alloc meter in verify/verify_and_update.